lib/ansible/plugins/connection/guestfs.py

```diff
@@ -0,0 +1,209 @@
+# Based on local.py (c) 2012, Michael DeHaan <michael.dehaan@gmail.com>
+# (c) 2013, Maykel Moya <mmoya@speedyrails.com>
+# (c) 2015, Toshio Kuratomi <tkuratomi@ansible.com>
+# (c) 2015, Michael Scherer <mscherer@redhat.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import os.path
+import guestfs
+import json
+import re
+
+from ansible.errors import AnsibleError
+from ansible.plugins.connection import ConnectionBase
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+# keep that code python2 and python3 compliant
+# we need to set $HOME so we are not left with
+# .ansible directory on /
+# also, I kept it {} free for clarity, hence the use of
+# disct instead of the short form with brackets.
+EXECUTION_SCRIPT = """
+import sys
+import json
+import os
+from subprocess import PIPE, Popen
+
+os.environ['HOME'] = '{}'
+
+cmd_file = sys.argv[1]
+if not os.path.exists(cmd_file):
+    sys.exit(1)
+
+cmd = json.loads(open(cmd_file).read())
+
+r = Popen(cmd, stdout=PIPE, stderr=PIPE, shell=True)
+(stdout, stderr) = r.communicate()
+
+output = dict()
+output['rc'] = r.returncode
+output['stdout'] = stdout.decode('UTF-8')
+output['stderr'] = stderr.decode('UTF-8')
+
+json.dump(output, open(cmd_file + '.out', 'w'))
+"""
+
+
+class Connection(ConnectionBase):
+    ''' Libguestfs based connections '''
+
+    transport = 'guestfs'
+    # not sure how it would react, so better disable it
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin,
+                                         *args, **kwargs)
+
+        self.disk = self._play_context.remote_addr
+
+        if not os.path.isfile(self.disk):
+            raise AnsibleError("%s is not a file, aborting" % self.disk)
+
+        self.guestfs = guestfs.GuestFS(python_return_dict=True)
+
+        self.guestfs.add_drive_opts(self.disk)
+
+    def _find_python(self):
```

Comment:
So, we need autodetection for Fedora and Ubuntu (since they come with python 3 pre installed, so I need to detect them for running raw to install python2). Now, I am not sure how should the plugin react if the user give a explicit python_interpreter, and if we can assume to be installed if given (especially for something like raw module, which should work without python).

In fact, I wonder if I shouldn't also have a wrapper in shell, since you can't use it on freebsd image with the current system of using a python wrapper if python is not installed (which is the case on freebsd among others). And I might need it soon for netbsd images.